File: compiler/feedback/bounds.go
package feedback

import (
	"wipple/database"
	"wipple/queries"
	"wipple/typecheck"
)

func registerBounds() {
	register(Feedback[typecheck.ResolvedBound]{
		Id:    "unresolved-bound",
		Rank:  RankBounds,
		Query: queries.UnresolvedBound,
		Render: func(render *Render, node database.Node, bound typecheck.ResolvedBound) {
			render.WriteNode(node)
			render.WriteString(" requires the instance ")
			render.WriteBound(bound)
			render.WriteString(", but this instance isn't defined.")
			render.WriteBreak()
			render.WriteString("Double-check that these types are correct.")
		},
	})

	type errorInstanceData struct {
		bound    typecheck.ResolvedBound
		group    *typecheck.Group
		comments queries.CommentsData
		trace    []typecheck.Constraint
	}

	register(Feedback[errorInstanceData]{
		Id:   "error-instance",
		Rank: RankCustom,
		Query: func(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(data errorInstanceData)) {
			typed, ok := database.GetFact[typecheck.TypedFact](node)
			if !ok || typed.Group == nil {
				return
			}

			queries.ErrorInstance(db, node, filter, func(bound typecheck.ResolvedBound, comments queries.CommentsData, trace []typecheck.Constraint) {
				f(errorInstanceData{
					bound:    bound,
					group:    typed.Group,
					comments: comments,
					trace:    trace,
				})
			})
		},
		On: func(data errorInstanceData) []database.Node {
			var nodes []database.Node

			// If the error instance links to a single node, put that node
			// first...
			if len(data.comments.Links) == 1 {
				for _, link := range data.comments.Links {
					nodes = append(nodes, link.Node)
				}
			} else {
				// ...otherwise, put the source first
				nodes = append(nodes, data.bound.Source)

				for _, link := range data.comments.Links {
					nodes = append(nodes, link.Node)
				}
			}

			nodes = append(nodes, data.comments.Nodes...)
			nodes = append(nodes, data.group.Nodes...)

			database.SortByProximity(nodes, data.bound.Source)

			return nodes
		},
		Render: func(render *Render, node database.Node, data errorInstanceData) {
			render.WriteComments(data.comments)

			if len(data.trace) > 0 {
				render.WriteBreak()
				seen := map[database.Node]struct{}{}
				for _, constraint := range data.trace {
					render.WriteConstraint("\n\n  -  ", constraint, seen)
				}
			}
		},
	})
}

File: compiler/src/feedback/items/bounds.rs
use crate::{
    feedback::{FeedbackCtx, FeedbackRank, RegisteredFeedback},
    queries::{self, QueriedComments},
    typecheck::{Bound, Constraint, Group, Typed},
};
use std::{collections::BTreeSet, sync::Arc};

pub fn register(ctx: &mut FeedbackCtx) {
    ctx.register(RegisteredFeedback::new(
        "unresolved-bound",
        FeedbackRank::Bounds,
        queries::unresolved_bound,
        |bound| (bound.source_node.clone(), BTreeSet::new()),
        |w, bound| {
            w.write_node(&bound.source_node);
            w.write_string(" requires the instance ");
            w.write_bound(bound);
            w.write_string(", but this instance isn't defined.");
            w.write_break();
            w.write_string("Double-check that these types are correct.");
        },
    ));

    #[derive(Debug)]
    struct ErrorInstanceData {
        bound: Bound,
        group: Arc<Group>,
        comments: QueriedComments,
        trace: Vec<Box<dyn Constraint>>,
    }

    ctx.register(
        RegisteredFeedback::new(
            "error-instance",
            FeedbackRank::Custom,
            |ctx, f| {
                let Some(Typed { group: Some(group) }) = ctx.db.get(&ctx.node) else {
                    return;
                };

                queries::error_instance(ctx, &mut |data| {
                    f(ErrorInstanceData {
                        bound: data.bound,
                        group: group.clone(),
                        comments: data.comments,
                        trace: data.trace,
                    });
                });
            },
            |data| {
                let primary = data.bound.source_node.clone();

                let related = data
                    .comments
                    .links
                    .values()
                    .flat_map(|link| [&link.node].into_iter().chain(&link.related))
                    .chain(&data.comments.nodes)
                    .chain(&data.group.nodes)
                    .cloned()
                    .collect::<BTreeSet<_>>();

                (primary, related)
            },
            |w, data| {
                w.write_comments(&data.comments);

                if !data.trace.is_empty() {
                    w.write_break();
                    for constraint in &data.trace {
                        w.write_constraint("\n\n  -  ", constraint.as_ref());
                    }
                }
            },
        )
        .show_graph(),
    );
}
